torch/_dynamo/guards.py

```diff
@@ -468,11 +469,54 @@ def TENSOR_MATCH(self, guard: Guard):
                 for term in terms:
                     real_value = self.get(tensor_name + "." + term)
                     code.append(f"{tensor_name}.{term} == {real_value}")
-                self._produce_guard_code(guard, code)
             else:
                 self.tensor_check_names.append(tensor_name)
                 self.tensor_check_examples.append(value)
 
+            # A frame is valid for reuse with dynamic dimensions if the new dynamic dimensions are a
+            # strict subset of the old.
+            #
+            # The logic here is as follows:
+            #
+            # Every mark_dynamic directive is a user-knows-best command, which can incur a raise at tracing
+            # time if we find guards that run counter to the user directive.
+            #
+            # If the frame is compiled with any marked dynamic indices, let's call that set of indices X.
+            # When we evaluated inputs against the guards, given the same tensor with potentially new dynamic indices, let's call that set Y.
+            # Y must be a strict subset of X.
+            #
+            # This is the case because any newly introduced mark_dynamic directives have a chance of
+            # incurring raises, failing compilation. Any existing mark_dynamic indices that we lost are safe to lose
+            # as all it means is that we have gotten rid of a user directive which could incur a raise at compile time.
+            # In the case of when there is no Y, that is, there are no dynamic indices marked at all, the frame is safe to reuse
+            # as an empty set is a safe degeneration - that is, a strictly static tensor is always valid for a frame compiled with that same
+            # tensor + more onerous user directives.
+            static, message = tensor_shape_should_be_static(
+                value, guard.source, is_tensor=True
+            )
```

Comment:
It's important to know that this calculation does NOT depend on `_dynamo_dynamic_indices`. If it did, this condition would be wrong. But it's easy to get the wrong idea, because isn't a tensor shape that doesn't have dynamic indices a static tensor? A rename might help here: `tensor_always_has_static_shape` or something like that, letting us know that it's not related to user annots.

Reply:
Sure